src/java.base/share/classes/sun/net/www/protocol/jar/JarURLConnection.java

```diff
@@ -88,8 +89,18 @@ public void close () throws IOException {
             try {
                 super.close();
             } finally {
-                if (!getUseCaches()) {
-                    jarFile.close();
+                try {
+                    if (!getUseCaches()) {
+                        jarFile.close();
+                    }
+                } finally {
+                    if (jarFileURLConnection instanceof FileURLConnection) {
+                        FileURLConnection fileURLConnection =
+                            (FileURLConnection) jarFileURLConnection;
+                        if (fileURLConnection.isConnected()) {
+                            fileURLConnection.getInputStream().close();
+                        }
+                    }
```

Comment:
This can use `instanceof` pattern matching:
```suggestion
                    if (jarFileURLConnection instanceof FileURLConnection fileURLConnection
                        && fileURLConnection.isConnected()) {
                        fileURLConnection.getInputStream().close();
                    }
```